stdlib/LinearAlgebra/src/qr.jl

```diff
@@ -527,21 +527,30 @@ AbstractMatrix{T}(Q::QRPackedQ) where {T} = QRPackedQ{T}(Q)
 QRCompactWYQ{S}(Q::QRCompactWYQ) where {S} = QRCompactWYQ(convert(AbstractMatrix{S}, Q.factors), convert(AbstractMatrix{S}, Q.T))
 AbstractMatrix{S}(Q::QRCompactWYQ{S}) where {S} = Q
 AbstractMatrix{S}(Q::QRCompactWYQ) where {S} = QRCompactWYQ{S}(Q)
-Matrix(A::AbstractQ{T}) where {T} = lmul!(A, Matrix{T}(I, size(A.factors, 1), min(size(A.factors)...)))
-Array(A::AbstractQ) = Matrix(A)
-
-size(A::Union{QR,QRCompactWY,QRPivoted}, dim::Integer) = size(getfield(A, :factors), dim)
-size(A::Union{QR,QRCompactWY,QRPivoted}) = size(getfield(A, :factors))
-size(A::AbstractQ, dim::Integer) = 0 < dim ? (dim <= 2 ? size(getfield(A, :factors), 1) : 1) : throw(BoundsError())
-size(A::AbstractQ) = size(A, 1), size(A, 2)
-
+Matrix(Q::Union{QRCompactWYQ,QRPackedQ}) where {T} = lmul!(A, Matrix{T}(I, size(Q, 1), min(size(A.factors)...)))
+Array(Q::AbstractQ) = Matrix(Q)
+
+size(F::Union{QR,QRCompactWY,QRPivoted}, dim::Integer) = size(getfield(F, :factors), dim)
+size(F::Union{QR,QRCompactWY,QRPivoted}) = size(getfield(F, :factors))
+function size(Q::Union{QRCompactWYQ,QRPackedQ}, dim::Integer)
+    if 0 < dim
+        if dim <= 2
+            return size(getfield(Q, :factors), 1)
+        else
+            return 1
+        end
+    else
+        throw(BoundsError())
+    end
+end
```

Comment:
Seems easier to define this as:
```jl
size(Q::Union{QRCompactWYQ,QRPackedQ}, dim::Integer) = size(getfield(Q, :factors), dim == 2 ? 1 : dim)
```

Reply:
Are there cases possible, where Q is not a square? So why not
```
size(A::AbstractQ, dim::Integer) = 0 < dim ? (dim <= 2 ? size(getfield(A, :factors), 1) : 1) : throw(BoundsError())
```
was correct in my opinion.
The definition of `Matrix` could make use of `size` in this way:
```
Matrix(Q::AbstractQ{T}) where {T} = lmul!(Q, Matrix{T}(I, size(Q, 1), size(Q, 2)))
```
The changes in `SuiteSparse` are not needed in this case.

Reply:
Line 530 now produces a runtime error in the dense case. Also test case is missing to reveal that.

Reply:
It's unclear what "Line 530 now produces a runtime error in the dense case." means. If you have a specific example that fails then please post it.

Reply:
Regarding your first comment then Steve's version gives the same result as yours and mine. Steve's just reuses as much of `size(Matrix)` as possible which is nicer that my two if statements.

Reply:
The problem in line 530 is, that `T` in `Matrx{T}` is not defined:
```
julia> A = [1 2 1 1 1 1 ; 1 2 0 1 1 1; 1 2 0 1 1 1; 0 0 0 0 0 0];

julia> F = qrfact((Float64.(A)), Val(true))
QRPivoted{Float64,Array{Float64,2}}
Q factor:
4×4 LinearAlgebra.QRPackedQ{Float64,Array{Float64,2}}:
 -0.57735   0.816497  -8.75605e-17  0.0
 -0.57735  -0.408248  -0.707107     0.0
 -0.57735  -0.408248   0.707107     0.0
  0.0       0.0        0.0          1.0
R factor:
4×6 Array{Float64,2}:
 -3.4641  -0.57735   -1.73205  -1.73205  -1.73205  -1.73205
  0.0      0.816497   0.0       0.0       0.0       0.0    
  0.0      0.0        0.0       0.0       0.0       0.0    
  0.0      0.0        0.0       0.0       0.0       0.0    
permutation:
6-element Array{Int64,1}:
 2
 3
 1
 4
 5
 6

julia> Matrix(F.Q)
ERROR: UndefVarError: T not defined
Stacktrace:
 [1] Array{T,2} where T(::LinearAlgebra.QRPackedQ{Float64,Array{Float64,2}}) at /home/julia/julia/usr/share/julia/site/v0.7/LinearAlgebra/src/qr.jl:530
 [2] top-level scope

julia> VERSION
v"0.7.0-DEV.4516"

julia> versioninfo()
Julia Version 0.7.0-DEV.4516
Commit 1886be7a83 (2018-03-09 14:38 UTC)
Platform Info:
  OS: Linux (x86_64-redhat-linux)
  CPU: Intel(R) Core(TM) i7-3610QM CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, ivybridge)
Environment:
```

Reply:
Thanks.

> Also test case is missing to reveal that.

The tests did reveal this though.